File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return 'N'
        res = []
        queue = deque([root])
        while queue:
            node = queue.popleft()
            if not node:
                res.append('N')
            else:
                res.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
            
        return ','.join(res)

        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        nodeList = data.split(',')
        if nodeList[0] == 'N':
            return None
        root = TreeNode(int(nodeList[0]))

        queue = deque([root])
        index = 1
        while queue:
            node = queue.popleft()
            if nodeList[index] != 'N':
                node.left = TreeNode(int(nodeList[index]))
                queue.append(node.left)
            index += 1
            if nodeList[index] != 'N':
                node.right = TreeNode(int(nodeList[index]))
                queue.append(node.right)
            index += 1
        return root

File: Data Structures & Algorithms/serialize-and-deserialize-n-ary-tree/submission-2.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Codec {
    // Encodes a tree to a single string.
    public String serialize(Node root) {
        if (root == null) return "N";
        String encoding = "";
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            Node cur = q.remove();
            if (cur != null) {
                int size = (cur.children != null) ? cur.children.size() : 0;
                encoding += cur.val + "#" + size + ",";
                for (Node child : cur.children) {
                    q.add(child);
                }
            } else {
                encoding += "N,";
            }
        }
        return encoding;
    }
	
    // Decodes your encoded data to tree.
    public Node deserialize(String data) {
        if (data.equals("N")) return null;
        String[] nodes = data.split(",");
        String[] rootComponents = nodes[0].split("#");
        int rootVal = Integer.parseInt(rootComponents[0]);
        int rootChildrenSize = Integer.parseInt(rootComponents[1]);
        Queue<Pair<Node, Integer>> q = new LinkedList<>();
        Node root = new Node(rootVal);
        if (rootChildrenSize != 0) {
            root.children = new ArrayList<>();
        }

        q.add(new Pair<>(root, rootChildrenSize));
        int i = 1;
        while (!q.isEmpty() && i < nodes.length) {
            Pair<Node, Integer> cur = q.remove();
            Node parent = cur.getKey();
            int numChildren = cur.getValue();

            for (int j = 0; j < numChildren; j++) {
                String[] childParts = nodes[i++].split("#");
                int childVal = Integer.parseInt(childParts[0]);
                int childCount = Integer.parseInt(childParts[1]);
                Node childNode = new Node(childVal);
                if (childCount != 0) {
                    childNode.children = new ArrayList<>();
                }
                parent.children.add(childNode);
                q.add(new Pair<>(childNode, childCount));
            }
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
